include/picongpu/plugins/randomizedParticleMerger/VoronoiCell.hpp

```diff
@@ -0,0 +1,371 @@
+            cupla::atomicAdd( acc, &this->numRealParticles, weighting, ::alpaka::hierarchy::Threads{} );
+
+            const floatD_X position2 = position * position;
+
+            for( int i = 0; i < simDim; i++ )
+            {
+                cupla::atomicAdd( acc, &this->meanPositionValue[i], weighting * position[i], ::alpaka::hierarchy::Threads{} );
+                cupla::atomicAdd( acc, &this->meanPositionSquaredValue[i], weighting * position2[i], ::alpaka::hierarchy::Threads{} );
+            }
+
+            const float3_X momentum2 = momentum * momentum;
+
+            for( int i = 0; i < DIM3; i++ )
+            {
+                cupla::atomicAdd( acc, &this->meanMomentumValue[i], weighting * momentum[i], ::alpaka::hierarchy::Threads{} );
+                cupla::atomicAdd( acc, &this->meanMomentumSquaredValue[i], weighting * momentum2[i], ::alpaka::hierarchy::Threads{} );
+            }
+        }
+
+        /** Counting parameters that are necessary before processing vornoi cell:
+         *  mean values and expected number of macro particles
+         *
+         * @param minMacroParticlesToDivide min number of macroparticles in a cell
+         *                                  such that the cell is always subdivided
+         * @param ratioKeptParticles ratio of particles that are kept on average
+         */
+        HDINLINE
+        void finalizePrecalculationValues(
+            const uint32_t minMacroParticlesToDivide,
+            const float_X ratioKeptParticles
+        )
+        {
+            finalizeMeanValues();
+            finalizeExpectedNumberParticles( minMacroParticlesToDivide, ratioKeptParticles );
+
+        }
+
+        //! Finalize calculation of mean values
+        HDINLINE
+        void finalizeMeanValues()
+        {
+            meanMomentumValue /= numRealParticles;
+            meanPositionValue /= numRealParticles;
+            meanMomentumSquaredValue /= numRealParticles;
+            meanPositionSquaredValue /= numRealParticles;
+        }
+
+        /** Count expected number of particles in the cell
+         *
+         * @param minMacroParticlesToDivide min number of macroparticles in a cell
+         *                                  such that the cell is always subdivided
+         * @param ratioKeptParticles ratio of particles that are kept on average
+         */
+        HDINLINE
+        void finalizeExpectedNumberParticles(
+            uint32_t const minMacroParticlesToDivide,
+            float_X const ratioKeptParticles
+        )
+        {
+            // Special case for the original voronoi cells
+            if ( parentExpectedNumMacroParticles < 0 )
+            {
+                expectedNumMacroParticles = numMacroParticles * ratioKeptParticles;
+                return;
+            }
+
+            // Algorithm stop conditions for 1 and 2 macroparticles
+            if ( numMacroParticles == 1 )
+                expectedNumMacroParticles = 1;
```

Comment:
should be `1.0_X` because expectedNumMacroParticles is of type `float_X`

Please also check the other lines in the if cascade.